src/bio/bio_xstream.c

```diff
@@ -160,6 +245,18 @@ populate_whitelist(struct spdk_env_opts *opts)
 			break;
 		}
 
+		if (vmd_enabled) {
+			if (strncmp(trid->traddr, "0", 1) != 0) {
+				/*
+				 * We can assume this is the transport id of the
+				 * backing NVMe SSD behind the VMD. DPDK will
+				 * not recognize this transport ID, instead need
+				 * to pass VMD address as the whitelist param.
+				 */
+				traddr_to_vmd(trid->traddr, trid->traddr);
```

Comment:
@sydneyvanda might have different ideas as she wrote this but my take on it is as follows:
- We can specify the fully qualified backing device ssd pci addresses (that exist behind vmd addresses) e.g. `5d0505:01:00.1` in nvme.conf because spdk understands having the vmd concatenated address in the domain field of the backing device address (this is the way vmd has been implemented in spdk). 
- This PCI address format will not be understood by dpdk as it's not a standard format, therefore , when we pass the whitelist from our spdk app to dpdk env we need to pass the vmd address rather than the backing device addresses. (just implemented logic to avoid duplicate entries in white list after that processing)

One possible way of reducing string manipulation in bio is to see if supplying the vmd address in addition to the backing addresses in nvme.conf is possible, then we could just pass any PCI addresses with blank domain fields in the dpdk whitelist.

Reply:
From a maintenance and testing perspective, it seems better to handle string manipulation in Go rather than C. I suspect that this address translation logic will continue to evolve as the feature gets more use and encounters more real-world scenarios. Adding unit tests for different address formats is trivial in Go, less so in C.

Reply:
I agree this would be better to handle in Go, but currently not sure how to accomplish this. Seems like this solution will work at least for the initial VMD implementation